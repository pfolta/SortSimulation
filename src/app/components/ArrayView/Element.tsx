import React from "react";
import Measure from "react-measure";
import styled from "styled-components";

interface ElementProps {
    array: number[];
    index: number;
    maxValue: number;
}

interface StyledElementProps extends ElementProps {
    contentWidth?: number;
}

const StyledElement = styled.div.attrs<StyledElementProps>(({ array, index, maxValue, contentWidth }) => ({
    style: {
        fontSize: computeFontSize(contentWidth) + "px",
        height: (array[index] / maxValue) * 100 + "%"
    }
}))<StyledElementProps>`
    background-color: #4c566a;
    color: white;

    overflow: hidden;

    font-size: 0;
    text-align: center;

    :hover {
        background-color: #ebcb8b;
    }
`;

const computeFontSize = (width?: number): number => {
    const size = width ? width / 2 : 0;
    return size < 8 ? 0 : size;
};

const Element = (props: ElementProps): JSX.Element => (
    <Measure>
        {({ contentRect, measureRef }) => (
            <StyledElement ref={measureRef} {...props} contentWidth={contentRect.entry?.width}>
                {props.array[props.index]}
            </StyledElement>
        )}
    </Measure>
);

export default Element;
export { StyledElement };
export { ElementProps };
