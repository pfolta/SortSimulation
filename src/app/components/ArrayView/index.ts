export { default as ArrayView, ArrayViewProps } from "@/app/components/ArrayView/ArrayView";
export { default as Element, ElementProps, StyledElement } from "@/app/components/ArrayView/Element";
