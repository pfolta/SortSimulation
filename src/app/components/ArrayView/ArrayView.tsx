import React, { useMemo } from "react";
import styled from "styled-components";

import { Element } from "@/app/components/ArrayView";

interface ArrayViewProps {
    array: number[];
}

const Wrapper = styled.div<ArrayViewProps>`
    display: grid;
    grid-template-columns: repeat(${(props) => props.array.length}, 1fr);
    column-gap: 1px;
    align-items: end;

    height: 30vh;
`;

const ArrayView = ({ array }: ArrayViewProps): JSX.Element => {
    const maxValue = useMemo(() => Math.max(...array), [array]);

    return (
        <Wrapper array={array}>
            {array.map((_, index) => (
                <Element key={index} array={array} index={index} maxValue={maxValue} />
            ))}
        </Wrapper>
    );
};

export default ArrayView;
export { ArrayViewProps };
