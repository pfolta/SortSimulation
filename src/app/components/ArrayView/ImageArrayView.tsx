import React, { memo } from "react";

import { ArrayView, ArrayViewProps } from "@/app/components/ArrayView";
import { ImageElement } from "@/app/components/ArrayViewElement";

interface ImageArrayViewProps extends ArrayViewProps {
    grayscaleMap: boolean[];
    image: string;
}

const ImageArrayView = ({ array, grayscaleMap, image }: ImageArrayViewProps): JSX.Element => (
    <ArrayView array={array}>
        {array.map((value, index) => (
            <ImageElement key={index} value={value} size={array.length} image={image} grayscale={grayscaleMap[index]} />
        ))}
    </ArrayView>
);

export default memo(ImageArrayView);
export { ImageArrayViewProps };
