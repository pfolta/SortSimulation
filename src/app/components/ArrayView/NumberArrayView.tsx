import React, { memo } from "react";
import styled from "styled-components";

import { ArrayView, ArrayViewProps } from "@/app/components/ArrayView";
import { NumberElement } from "@/app/components/ArrayViewElement";

const StyledNumberArrayView = styled(ArrayView)`
    column-gap: 1px;
`;

const NumberArrayView = ({ array }: ArrayViewProps): JSX.Element => {
    const maxValue = Math.max(...array);

    return (
        <StyledNumberArrayView array={array}>
            {array.map((value, index) => (
                <NumberElement key={index} value={value} maxValue={maxValue} />
            ))}
        </StyledNumberArrayView>
    );
};

export default memo(NumberArrayView);
