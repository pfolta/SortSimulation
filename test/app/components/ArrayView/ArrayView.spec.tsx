import React from "react";
import { create } from "react-test-renderer";

import ArrayView from "@/app/components/ArrayView/ArrayView";
import { StyledElement } from "@/app/components/ArrayView/Element";

describe("ArrayView", () => {
    const array = [4, 2, 5, 1, 3];

    it("renders correctly", () => {
        const component = create(<ArrayView array={array} />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders every array element in the correct order", () => {
        const component = create(<ArrayView array={array} />);
        const arrayElements = component.root.findAllByType(StyledElement);

        expect(arrayElements).toHaveLength(array.length);
        arrayElements.forEach((bar, index) => expect(bar.props.children).toBe(array[index]));
    });
});
