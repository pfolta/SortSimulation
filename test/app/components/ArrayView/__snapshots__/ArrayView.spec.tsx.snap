// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArrayView renders correctly 1`] = `
.c1 {
  background-color: #4c566a;
  color: white;
  overflow: hidden;
  font-size: 0;
  text-align: center;
}

.c1:hover {
  background-color: #ebcb8b;
}

.c0 {
  display: grid;
  grid-template-columns: repeat(5,1fr);
  -webkit-column-gap: 1px;
  column-gap: 1px;
  -webkit-align-items: end;
  -webkit-box-align: end;
  -ms-flex-align: end;
  align-items: end;
  height: 30vh;
}

<div
  className="c0"
>
  <div
    className="c1"
    style={
      Object {
        "fontSize": "0px",
        "height": "80%",
      }
    }
  >
    4
  </div>
  <div
    className="c1"
    style={
      Object {
        "fontSize": "0px",
        "height": "40%",
      }
    }
  >
    2
  </div>
  <div
    className="c1"
    style={
      Object {
        "fontSize": "0px",
        "height": "100%",
      }
    }
  >
    5
  </div>
  <div
    className="c1"
    style={
      Object {
        "fontSize": "0px",
        "height": "20%",
      }
    }
  >
    1
  </div>
  <div
    className="c1"
    style={
      Object {
        "fontSize": "0px",
        "height": "60%",
      }
    }
  >
    3
  </div>
</div>
`;
